### PR TITLE
[PWGHF] Add finding leading particle in correlatorLcHadrons

### PR DESCRIPTION
**Pull request by 1481014945**
Dear, experts
Sorry for bothering you again.
I would like to add a function to find the leading particle in correlatorLcHadrons. This is similar to what we have in correlatorD0Hadrons, which you can see through this link.
https://github.com/AliceO2Group/O2Physics/pull/4962
https://github.com/AliceO2Group/O2Physics/pull/7722

### CODE REVIEW

File: PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx [hunk 1]
```diff
@@ -254,6 +258,46 @@ struct HfCorrelatorLcHadrons {
...
+    for (auto const& track : tracks) {
+      if (std::abs(track.dcaXY()) >= 1. || std::abs(track.dcaZ()) >= 1.) {
+        continue;
+      }
+      if (track.pt() > leadingParticle.pt()) {
+        leadingParticle = track;
+      }
+    }
+    int leadingIndex = leadingParticle.globalIndex();
+    return leadingIndex;
```
Comment (vkucera) [inline PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx:275]:
```suggestion
    return leadingParticle.globalIndex();
```

File: PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx [hunk 2]
```diff
@@ -254,6 +258,46 @@ struct HfCorrelatorLcHadrons {
...
+      }
+      if ((std::abs(mcParticle.pdgCode()) != kElectron) && (std::abs(mcParticle.pdgCode()) != kMuonMinus) && (std::abs(mcParticle.pdgCode()) != kPiPlus) && (std::abs(mcParticle.pdgCode()) != kKPlus) && (std::abs(mcParticle.pdgCode()) != kProton)) {
+        continue;
+      }
+      if (mcParticle.pt() > leadingParticle.pt()) {
+        leadingParticle = mcParticle;
+      }
+    }
+    int leadingIndex = leadingParticle.globalIndex();
+    return leadingIndex;
```
Comment (vkucera) [inline PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx:298]:
```suggestion
    return leadingParticle.globalIndex();
```

File: PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx [hunk 3]
```diff
@@ -96,35 +98,54 @@ struct HfTaskCorrelationLcHadrons {
   Configurable<std::vector<double>> sidebandRightInner{"sidebandRightInner", std::vector<double>{vecSidebandRightInner}, "Inner values of right sideband vs Pt"};
   Configurable<std::vector<double>> sidebandRightOuter{"sidebandRightOuter", std::vector<double>{vecSidebandRightOuter}, "Outer values of right sideband vs Pt"};
   Configurable<std::vector<double>> efficiencyLc{"efficiencyLc", std::vector<double>{vecEfficiencyLc}, "Efficiency values for Lc "};
+  Configurable<bool> isTowardTransverseAway{"isTowardTransverseAway", false, "Divide into three regions: toward, transverse, and away"};
+  Configurable<double> leadingParticlePtMin{"leadingParticlePtMin", 0., "Min for leading particle pt"};
 
   using LcHadronPairFull = soa::Join<aod::LcHadronPair, aod::LcHadronRecoInfo>;
 
+  enum Region {
+    Default = 0, // 默认值
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx:107]: No non-English comments, please.

File: PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx [hunk 4]
```diff
@@ -148,8 +169,20 @@ struct HfTaskCorrelationLcHadrons {
...
+    } else if (deltaPhi > 2. * o2::constants::math::PI / 3. && deltaPhi < 4. * o2::constants::math::PI / 3.) {
+      return Away;
+    } else {
+      return Transverse;
+    }
+  }
+
   void processData(LcHadronPairFull const& pairEntries)
   {
+    std::cout << "  ";
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx:185]:
```suggestion
```

File: PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx [hunk 5]
```diff
@@ -148,8 +169,20 @@ struct HfTaskCorrelationLcHadrons {
...
+  Region getRegion(double deltaPhi)
+  {
+    if (std::abs(deltaPhi) < o2::constants::math::PI / 3.) {
+      return Toward;
+    } else if (deltaPhi > 2. * o2::constants::math::PI / 3. && deltaPhi < 4. * o2::constants::math::PI / 3.) {
+      return Away;
+    } else {
+      return Transverse;
+    }
+  }
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx:181]: This function is already defined in `taskCorrelationD0Hadrons.cxx`. Please move it out into a common header.

**Comment by vkucera** [comment]
@1481014945 Please use proper Git commits with meaningful titles for your development. It is also not recommended to develop in the `master` branch. If you are still developing, please do not mark the PR as ready for review and keep it as a draft until you have the final version.

File: PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx (hunk 5 above)
Comment (vkucera) [reply PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx:181]: Same for the definition of `Region`.

Comment (1481014945) [reply PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx:181]: Thank you for your suggestion; I’ll address it as soon as possible. Also, regarding the getRegion function, should I create a new header file, or should I include it in an existing one?

Comment (vkucera) [reply PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx:181]: I would suggest a new header `PWGHF/HFC/Utils/utilsCorrelations.h`

Comment (1481014945) [reply PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx]: I have tried to add it in a new header PWGHF/HFC/Utils/utilsCorrelations.h, but I am not sure if it meets your suggestions.

Comment (vkucera) [reply PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx]: Thanks, it looks good.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file utilsCorrelations.h
+/// \brief Utilities for HFC analyses
+
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:14]: Add yourself as author.
```suggestion
/// \brief Utilities for HFC analyses
/// \author ...

```

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+#ifndef PWGHF_HFC_UTILS_UTILSCORRELATIONS_H_
+#define PWGHF_HFC_UTILS_UTILSCORRELATIONS_H_
+
+#include <TString.h>
+#include <cmath>
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:19]:
```suggestion
#include <cmath>
#include <TString.h>
```
Include the header which defines `o2::constants::math::PI`.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+enum Region {
+  Default = 0,
+  Toward,
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:23]: Define all the content inside a namespace, e.g. `o2::analysis::hf_correlations`.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  Away,
+  Transverse
+};
+
+Region getRegion(double deltaPhi)
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:28]:
```suggestion
Region getRegion(double deltaPhi)
```
Replace `double` with a template.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx (hunk 5 above)
Comment (vkucera) [reply PWGHF/HFC/Tasks/taskCorrelationLcHadrons.cxx]: What do you mean?

File: PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx [hunk 10]
```diff
@@ -254,6 +258,44 @@ struct HfCorrelatorLcHadrons {
     corrBinning = {{binsZVtx, binsMultiplicity}, true};
   }
 
+  // Find Leading Particle
+  template <typename TTracks>
+  int findLeadingParticle(TTracks const& tracks)
```
Comment (vkucera) [inline PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx:263]: `findLeadingParticle` and `findLeadingParticleMcGen` are also duplications of methods in `PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx`.
Please move these in the same header as well.

Comment (1481014945) [reply PWGHF/HFC/TableProducer/correlatorLcHadrons.cxx]: I have now tried to add it in the same header, but I am not sure if it meets your suggestions.

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 7 above)
Comment (vkucera) [reply PWGHF/HFC/Utils/utilsCorrelations.h]: You also need to include `TPDGCode.h` for the named PDG codes.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      leadingParticle = track;
+    }
+  }
+  int leadingIndex = leadingParticle.globalIndex();
+  return leadingIndex;
+}
+
+// ======= Find Leading Particle for McGen ============
+template <typename TMcParticles>
+int findLeadingParticleMcGen(TMcParticles const& mcParticles, const float etaTrackMax, const float ptTrackMin)
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:60]: Replace `float` with templates.

Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h:60]: const float etaTrackMax, const float ptTrackMin; I don't know why I always get an error when using templates.

Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h]: I have completed it now.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      leadingParticle = track;
+    }
+  }
+  int leadingIndex = leadingParticle.globalIndex();
+  return leadingIndex;
+}
+
+// ======= Find Leading Particle for McGen ============
+template <typename TMcParticles, typename T1, typename T2>
+int findLeadingParticleMcGen(TMcParticles const& mcParticles, T1 etaTrackMax, T2 ptTrackMin)
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:60]:
```suggestion
int findLeadingParticleMcGen(TMcParticles const& mcParticles, T1 const etaTrackMax, T2 const ptTrackMin)
```

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  }
+}
+
+// Find Leading Particle
+template <typename TTracks>
+int findLeadingParticle(TTracks const& tracks)
+{
+  auto leadingParticle = tracks.begin();
+  for (auto const& track : tracks) {
+    if (std::abs(track.dcaXY()) >= 1. || std::abs(track.dcaZ()) >= 1.) {
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:47]: These should be configurable.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file utilsCorrelations.h
+/// \brief Xu Wang <waxu@cern.ch>
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:13]: Fix

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+#ifndef PWGHF_HFC_UTILS_UTILSCORRELATIONS_H_
+#define PWGHF_HFC_UTILS_UTILSCORRELATIONS_H_
+
+#include <cmath>
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:18]: Fix

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  Default = 0,
+  Toward,
+  Away,
+  Transverse
+};
+
+template <typename T>
+Region getRegion(T deltaPhi)
+{
+  if (std::abs(deltaPhi) < o2::constants::math::PI / 3.) {
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:32]:
```suggestion
  if (std::abs(deltaPhi) < o2::constants::math::PIThird) {
```

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    return Toward;
+  } else if (deltaPhi > 2. * o2::constants::math::PI / 3. && deltaPhi < 4. * o2::constants::math::PI / 3.) {
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:34]:
```suggestion
  } else if (deltaPhi > 2. * o2::constants::math::PIThird && deltaPhi < 4. * o2::constants::math::PIThird) {
```

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 12 above)
Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h]: Sorry,When I use const, I encounter this kind of error.
/home/wangxu/alice/sw/SOURCES/O2Physics/master/0/PWGHF/HFC/Utils/utilsCorrelations.h:63:38: error: no match for 'operator>' (operand types are 'float' and 'const o2::framework::Configurable<float>')
   63 |       if (std::abs(mcParticle.eta()) > etaTrackMax)

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 15 above)
Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h:18]: I'm sorry, but I don't quite understand what you mean by "fix."

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 12 above)
Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h:60]: When I added .value(), it still produced an error.
/home/wangxu/alice/sw/SOURCES/O2Physics/master/0/PWGHF/HFC/Utils/utilsCorrelations.h:63:57: error: expression cannot be used as a function
   63 |       if (std::abs(mcParticle.eta()) > etaTrackMax.value())

Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h:60]: I have referenced other people's code, which uses float.
 https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/Utils/utilsAnalysis.h#L49

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 15 above)
Comment (vkucera) [reply PWGHF/HFC/Utils/utilsCorrelations.h:18]: You have not addressed my previous comments correctly. See https://github.com/AliceO2Group/O2Physics/pull/8183#discussion_r1820613158

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 12 above)
Comment (vkucera) [reply PWGHF/HFC/Utils/utilsCorrelations.h:60]: You added `()` which is wrong.

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 15 above)
Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h:18]: Sorry, I overlooked this message.

Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h:18]: Sorry, I overlooked this message.

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 12 above)
Comment (1481014945) [reply PWGHF/HFC/Utils/utilsCorrelations.h]: You are very impressive, and I am truly grateful. I have solved it.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 18]
```diff
@@ -10,12 +10,13 @@
 // or submit itself to any jurisdiction.
 
 /// \file utilsCorrelations.h
-/// \brief Xu Wang <waxu@cern.ch>
+/// \author Xu Wang <waxu@cern.ch>
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:13]: You removed the correct `\brief` line. Please put it back.

File: PWGHF/HFC/Utils/utilsCorrelations.h (hunk 15 above)
Comment (vkucera) [reply PWGHF/HFC/Utils/utilsCorrelations.h:18]: The header with math constants is still missing.

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 19]
```diff
@@ -29,22 +30,22 @@ enum Region {
...
 // Find Leading Particle
-template <typename TTracks>
-int findLeadingParticle(TTracks const& tracks)
+template <typename TTracks, typename T1, typename T2>
+int findLeadingParticle(TTracks const& tracks, T1 const dcaXYTrackMax, T2 const dcaZTrackMax)
 {
   auto leadingParticle = tracks.begin();
   for (auto const& track : tracks) {
-    if (std::abs(track.dcaXY()) >= 1. || std::abs(track.dcaZ()) >= 1.) {
+    if (std::abs(track.dcaXY()) >= dcaXYTrackMax.value || std::abs(track.dcaZ()) >= dcaZTrackMax.value) {
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:48]: The `.value` should not be called inside the helper function. It should be called inside the task when calling the function.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/HFC/Utils/utilsCorrelations.h [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+/// \file utilsCorrelations.h
+/// \brief Utilities for HFC analyses
+/// \author Xu Wang <waxu@cern.ch>
+
+#ifndef PWGHF_HFC_UTILS_UTILSCORRELATIONS_H_
+#define PWGHF_HFC_UTILS_UTILSCORRELATIONS_H_
+
+#include <cmath>
+#include "CommonConstants/PhysicsConstants.h"
+#include <TPDGCode.h>
```
Comment (vkucera) [inline PWGHF/HFC/Utils/utilsCorrelations.h:21]: Please fix the includes according to the HF guidelines.

**Review by vkucera — CHANGES_REQUESTED** [review]

**Review by vkucera — APPROVED** [review]
Thanks @1481014945 for addressing my comments. It looks good to me.
@apalasciano @deepathoms please review.

**Comment by 1481014945** [comment]
@vkucera However, @apalasciano and @deepathoms haven't replied yet. What should be done next?

**Pull request merged**